scarb/src/core/registry/client/cache.rs

```diff
@@ -0,0 +1,75 @@
+use std::path::PathBuf;
+
+use anyhow::{bail, Result};
+use tracing::trace;
+
+use crate::core::registry::client::{BeforeNetworkCallback, RegistryClient, RegistryResource};
+use crate::core::registry::index::IndexRecords;
+use crate::core::{Config, ManifestDependency, PackageId};
+
+pub struct RegistryClientCache<'c> {
+    client: Box<dyn RegistryClient + 'c>,
+    _config: &'c Config,
+}
+
+impl<'c> RegistryClientCache<'c> {
+    pub fn new(client: Box<dyn RegistryClient + 'c>, config: &'c Config) -> Result<Self> {
+        Ok(Self {
+            client,
+            _config: config,
+        })
+    }
+
+    /// Layer over [`RegistryClient::get_records`] that caches the result.
+    ///
+    /// It takes [`ManifestDependency`] instead of [`PackageName`] to allow performing some
+    /// optimizations by pre-filtering index records on cache-level.
+    #[tracing::instrument(level = "trace", skip_all)]
+    pub async fn get_records_with_cache(
```

Comment:
Do we need `_with_cache` suffixes here?

Reply:
I wanted to be more explicit at usage place, especially as this method has a different signature than the one in `RegistryClient` trait.

Reply:
> has a different signature 
This is a strong point indeed